From json

arrrrny opened the pull request:
# Summary
Added JSON serialization support to all model classes in the `lib/common/models` directory.

Main idea is that, I want to be able to write against an interface that gets a Json config. This allows the dart web scraper to be injected on a datasource. Also it is convenient to save the parsers as json in a database.

## Changes Made
- ✅ **UrlCleaner**: Added JSON serialization with proper null handling
- ✅ **Config**: Added JSON serialization with enum handling for UserAgentDevice
- ✅ **Data**: Added JSON serialization with URI parsing
- ✅ **WebScraperError**: Added JSON serialization for error messages
- ✅ **Parser**: Added JSON serialization with enum handling (excluding non-serializable cleaner function)
- ✅ **UrlTarget**: Added JSON serialization with nested object handling
- ✅ **Optional**: Added comprehensive JSON serialization for all fields and enums

## Features Implemented
- **Full enum serialization/deserialization support** - All enums are properly converted to/from their string representations
- **Proper null safety handling** - Nullable fields are handled correctly with appropriate defaults
- **Type-safe collection conversions** - Lists and Maps are explicitly cast for safety
- **Nested object serialization** - Complex objects like UrlCleaner in UrlTarget are properly serialized
- **URI conversion support** - Uri objects are converted to/from strings seamlessly
- **Default value preservation** - Default values are maintained during deserialization

## Benefits
- **Configuration storage and retrieval** - Enables saving and loading scraper configurations
- **API communication** - Facilitates sending/receiving configurations via APIs
- **Testing and debugging** - Makes it easier to create test configurations and debug issues
- **Configuration management** - Allows for easier management of complex scraper setups

## Technical Notes
- The `CleanerFunction` typedef in the Parser model cannot be serialized to JSON as functions are not JSON-serializable (noted in code comments)
- All enum values are converted to their string names (without class prefix) for cleaner JSON representation
- Complex nested structures are properly handled with recursive serialization
- Comprehensive field coverage ensures all properties are serializable

All models now support complete JSON serialization and deserialization, making the library more versatile for storage, configuration management, and API integration.

sukhcha-in reviewed (commented):
How did you register a cleaner to use it with a cleaner name?

arrrrny commented:
Since I am storing my config map and parsers, It is very useful to store the cleaner function names as refence. This assure I can reuse common cleaners for all domains. 

```
import 'package:dart_web_scraper/dart_web_scraper.dart';

typedef CleanerFunction = Object? Function(Data data, bool debug);

/// Registry and utilities for mapping cleanerName to CleanerFunction.
class CleanerUtils {
  static final Map<String, CleanerFunction> _registry = {};

  /// Register a cleaner function with a string key.
  static void register(String name, CleanerFunction function) {
    _registry[name] = function;
  }

  /// Resolve a cleaner function by its name.
  static CleanerFunction? resolve(String? name) {
    if (name == null) return null;
    return _registry[name];
  }

  /// Helper to resolve the cleaner for a parser instance.
  /// If the parser has a direct cleaner, returns it.
  /// Otherwise, resolves by cleanerName from the registry.
  static CleanerFunction? resolveCleanerForParser(Parser parser) {
    if (parser.cleaner != null) return parser.cleaner;
    if (parser.cleanerName != null) return resolve(parser.cleanerName);
    return null;
  }
}
```


```
import 'cleaner_utils.dart';

class GenericCleaners {
  static void registerCleaners() {
    CleanerUtils.register(
        'price', (data, debug) => parsePrice(data.obj.toString()));
    CleanerUtils.register(
        'ratingScore', (data, debug) => parseRatingScore(data.obj.toString()));
    CleanerUtils.register(
        'reviewCount', (data, debug) => parseReviewCount(data.obj.toString()));
  }

  static double? parsePrice(String priceText) {
    if (priceText.isEmpty) return null;
    String cleaned = priceText.replaceAll(RegExp(r'[^\d,\.]'), '');
    if (cleaned.contains(',') && cleaned.contains('.')) {
      cleaned = cleaned.replaceAll('.', '').replaceAll(',', '.');
    } else if (cleaned.contains(',') && !cleaned.contains('.')) {
      cleaned = cleaned.replaceAll(',', '.');
    } else if (cleaned.contains('.') && !cleaned.contains(',')) {
      if (cleaned.split('.').last.length != 2) {
        cleaned = cleaned.replaceAll('.', '');
      }
    }
    return double.tryParse(cleaned);
  }

  static double? parseRatingScore(String ratingText) {
    if (ratingText.isEmpty) return null;
    String cleaned =
        ratingText.replaceAll(RegExp(r'[^\d,\.]'), '').replaceAll(',', '.');
    return double.tryParse(cleaned);
  }

  static int? parseReviewCount(String reviewCountText) {
    if (reviewCountText.isEmpty) return null;
    String cleaned = reviewCountText.replaceAll(RegExp(r'[^\d]'), '');
    return int.tryParse(cleaned);
  }
}
```
